### ISSUE TITLE
Unable to trigger PostConfirmation Cognito Event

### ISSUE THREAD

**Issue by rudresh4** [aws/aws-lambda-go#44]
Hello,

I'm trying to trigger a lambda written in Golang by Cognito event "PostConfirmation". It seems that the failure is caused by json conversion of Cognito event:

"PostConfirmation failed with error json: cannot unmarshal string into Go struct field CognitoEvent.version of type int."

Here's the Lambda that I'm trying to use:

```
package main

import (
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"log"
)

func Handler(request events.CognitoEvent) (string, error) {
	log.Print("Received: ", request.EventType)
	return "Success", nil
}

func main() {
	lambda.Start(Handler)
}
```

**Comment by adamo57** [comment]
what does the data that you are sending it look like?

**Comment by rudresh4** [comment]
To be clear, my lambda handler doesn't get invoked. This is failing before that, which means I can't add any logic to iterate through the event.

My assumption is that maybe Cognito is sending version payload in "string" but aws-lambda-go is expecting int? https://github.com/aws/aws-lambda-go/blob/master/events/cognito.go#L13

**Comment by rudresh4** [comment]
Post-confirmation parameters sent to Lambda from Cognito: https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-lambda-trigger-syntax-post-confirmation.html

**Comment by adamo57** [comment]
From the pre configured AWS test events for Cognito, the version should be passed as an int.

**Comment by rayhaanq** [comment]
Hey there, any update or workaround on this?

**Comment by dfalgout** [comment]
You could just define them yourself and change the type of Version from int to string:

```
type CognitoEvent struct {
	DatasetName    string                          `json:"datasetName"`
	DatasetRecords map[string]CognitoDatasetRecord `json:"datasetRecords"`
	EventType      string                          `json:"eventType"`
	IdentityID     string                          `json:"identityId"`
	IdentityPoolID string                          `json:"identityPoolId"`
	Region         string                          `json:"region"`
	Version        string                          `json:"version"`
}

// CognitoDatasetRecord represents a record from an AWS Cognito event
type CognitoDatasetRecord struct {
	NewValue string `json:"newValue"`
	OldValue string `json:"oldValue"`
	Op       string `json:"op"`
}

type Response struct {
	Message string `json:"message"`
}

func Handler(ctx context.Context, event CognitoEvent) (Response, error) {
	for datasetName, datasetRecord := range event.DatasetRecords {
		fmt.Printf("[%s -- %s] %s -> %s -> %s \n",
			event.EventType,
			datasetName,
			datasetRecord.OldValue,
			datasetRecord.Op,
			datasetRecord.NewValue)
	}

	return Response{
		Message: "Testing",
	}, nil
}

func main() {
	lambda.Start(Handler)
}
```

**Comment by rayhaanq** [comment]
@dfalgout thank you

**Comment by rayhaanq** [comment]
@dfalgout the lambda function executes but when trying to add a response i keep getting an error when clicking the verify link in the email
`Unrecognizable lambda output`
This is using the same code as above

**Comment by cmello** [comment]
Hi @rayhaanq, the response for the Custom Message event should contain the emailSubject, emailMessage and smsMessage attributes. Please see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-lambda-trigger-examples.html#aws-lambda-triggers-custom-message-example and https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-lambda-trigger-syntax-custom-message.html . Thanks!

**Comment by rayhaanq** [comment]
@cmello thanks for getting back to me. This is for the PostConfirmation event.  I've been at this for a few hours now and I've finally got it to work with a successful response. This is the code that makes it functional. There's probably other fields missing but I guess they're not completely necessary.

These are the two structs that are necessary. `CognitoEventRequest` is a modified version of `events.CognitoEvent`. This was necessary as most of the fields don't work at all. The `CognitoEventResponse` is necessary to get an actual proper response without any errors.
```
type CognitoEventRequest struct {
	DatasetName    string                          `json:"datasetName"`
	DatasetRecords map[string]CognitoDatasetRecord `json:"datasetRecords"`
	TriggerSource  string                          `json:"triggerSource"`
	IdentityID     string                          `json:"identityId"`
	UserPoolID     string                          `json:"userPoolId"`
	Region         string                          `json:"region"`
	Version        string                          `json:"version"`
	Request        interface{}                     `json:"request"`
	Username       string                          `json:"userName"`
}

type CognitoEventResponse struct {
	TriggerSource string   `json:"triggerSource"`
	Username      string   `json:"userName"`
	UserPoolID    string   `json:"userPoolId"`
	Region        string   `json:"region"`
	Version       int      `json:"version"`
	Response      struct{} `json:"response"`
}
```

my handler definition
`func Handler(ctx context.Context, cognitoEvent CognitoEventRequest) (interface{}, error)`

within the function somewhere it's necessary to create the response object. This is what worked for me
```
res := CognitoEventResponse{}
res.Version = 1
res.TriggerSource = cognitoEvent.TriggerSource
res.Region = cognitoEvent.Region
res.UserPoolID = cognitoEvent.UserPoolID
res.Username = cognitoEvent.Username
```

and finally `return res, nil` when you're done.

It seems as if triggers are completely broken currently. Also the documentation is quite bare for doing this in golang. There should also be relevant response structs and obviously CognitEvent needs updating. Thanks

**Comment by cmello** [comment]
Hi @rayhaanq, thank you very much for sharing your solution. Submitted #59 for code review.

**Comment by marchinram** [comment]
thanks @rayhaanq! based on your code I got a pre signup trigger working to auto confirm in go.  I agree the docs are quite bare for golang and it took a lot of trial and error.

**Comment by bmoffatt** [comment]
Merged #73

**Comment by brydavis** [comment]
For those who stumble upon this in future, here is the entire solution altogether... (Credit given to everyone above, thanks!)

	package main

	import (
		"context"
		"fmt"

		"github.com/aws/aws-lambda-go/lambda"
	)

	type CognitoEventRequest struct {
		DatasetName    string                          `json:"datasetName"`
		DatasetRecords map[string]CognitoDatasetRecord `json:"datasetRecords"`
		TriggerSource  string                          `json:"triggerSource"`
		IdentityID     string                          `json:"identityId"`
		UserPoolID     string                          `json:"userPoolId"`
		Region         string                          `json:"region"`
		Version        string                          `json:"version"`
		Request        interface{}                     `json:"request"`
		Username       string                          `json:"userName"`
	}

	type CognitoDatasetRecord struct {
		NewValue string `json:"newValue"`
		OldValue string `json:"oldValue"`
		Op       string `json:"op"`
	}

	type CognitoEventResponse struct {
		TriggerSource string   `json:"triggerSource"`
		Username      string   `json:"userName"`
		UserPoolID    string   `json:"userPoolId"`
		Region        string   `json:"region"`
		Version       int      `json:"version"`
		Response      struct{} `json:"response"`
	}

	func Handler(ctx context.Context, cognitoEvent CognitoEventRequest) (interface{}, error) {
		for datasetName, datasetRecord := range cognitoEvent.DatasetRecords {
			fmt.Printf("[%s] %s -> %s -> %s \n",
				datasetName,
				datasetRecord.OldValue,
				datasetRecord.Op,
				datasetRecord.NewValue)
		}

		res := CognitoEventResponse{}
		res.Version = 1
		res.TriggerSource = cognitoEvent.TriggerSource
		res.Region = cognitoEvent.Region
		res.UserPoolID = cognitoEvent.UserPoolID
		res.Username = cognitoEvent.Username

		return res, nil
	}

	func main() {
		lambda.Start(Handler)
	}
